NullPointerException in LazyDeserializer due to error in double-checked locking code

The `unwrap()` method may return a null deserialiser if two threads access the method simultaneously. 

https://github.com/elastic/elasticsearch-java/blob/3e6f05c501da156062d02403d90ae7ce92d4a764/java-client/src/main/java/co/elastic/clients/json/LazyDeserializer.java#L41-L53

Stacktrace:

```
Caused by: java.lang.NullPointerException: null
	at co.elastic.clients.json.DelegatingDeserializer$SameType.deserialize(DelegatingDeserializer.java:43) ~[elasticsearch-java-7.16.0.jar:na]
	at co.elastic.clients.json.ObjectDeserializer$FieldObjectDeserializer.deserialize(ObjectDeserializer.java:72) ~[elasticsearch-java-7.16.0.jar:na]
	at co.elastic.clients.json.ObjectDeserializer.deserialize(ObjectDeserializer.java:176) ~[elasticsearch-java-7.16.0.jar:na]
	at co.elastic.clients.json.ObjectDeserializer.deserialize(ObjectDeserializer.java:137) ~[elasticsearch-java-7.16.0.jar:na]
	at co.elastic.clients.json.JsonpDeserializer.deserialize(JsonpDeserializer.java:75) ~[elasticsearch-java-7.16.0.jar:na]
	at co.elastic.clients.json.ObjectBuilderDeserializer.deserialize(ObjectBuilderDeserializer.java:79) ~[elasticsearch-java-7.16.0.jar:na]
	at co.elastic.clients.json.DelegatingDeserializer$SameType.deserialize(DelegatingDeserializer.java:43) ~[elasticsearch-java-7.16.0.jar:na]
	at co.elastic.clients.transport.endpoints.EndpointWithResponseMapperAttr$1.deserialize(EndpointWithResponseMapperAttr.java:56) ~[elasticsearch-java-7.16.0.jar:na]
	at co.elastic.clients.transport.rest_client.RestClientTransport.decodeResponse(RestClientTransport.java:325) ~[elasticsearch-java-7.16.0.jar:na]
	at co.elastic.clients.transport.rest_client.RestClientTransport.getHighLevelResponse(RestClientTransport.java:291) ~[elasticsearch-java-7.16.0.jar:na]
	at co.elastic.clients.transport.rest_client.RestClientTransport.performRequest(RestClientTransport.java:144) ~[elasticsearch-java-7.16.0.jar:na]
	at co.elastic.clients.elasticsearch.ElasticsearchClient.search(ElasticsearchClient.java:1487) ~[elasticsearch-java-7.16.0.jar:na]
```

Debugging screenshot where you can see `deserializer` is set, but `d` is `null`:
![image](https://user-images.githubusercontent.com/3715975/145815117-9aa248bc-417a-4e95-b69f-a5d6b34b90e8.png)

Looking at [the article](https://wiki.sei.cmu.edu/confluence/x/6zdGBQ) linked to in the source code, it looks like the "Compliant Solution (Immutable)" is desired. In this case, there is one assignment (`d = deserializer`) missing just before the null check inside the `synchonized` block.

So the code below should work better, making sure `d` is getting assigned with `deserializer`, for the cases where it has been assigned by another thread after the current thread has entered the synchronized block:

```
    protected JsonpDeserializer<T> unwrap() {
        // See SEI CERT LCK10-J https://wiki.sei.cmu.edu/confluence/x/6zdGBQ
        JsonpDeserializer<T> d = deserializer;
        if (d == null) {
            synchronized (this) {
                d = deserializer;
                if (d == null) {
                    d = ctor.get();
                    deserializer = d;
                }
            }
        }
        return d;
    }
```

Version: 7.16.0

--- a/java-client/src/main/java/co/elastic/clients/json/LazyDeserializer.java
+++ b/java-client/src/main/java/co/elastic/clients/json/LazyDeserializer.java
@@ -41,14 +41,15 @@ class LazyDeserializer<T> extends DelegatingDeserializer.SameType<T> {
     protected JsonpDeserializer<T> unwrap() {
         // See SEI CERT LCK10-J https://wiki.sei.cmu.edu/confluence/x/6zdGBQ
         JsonpDeserializer<T> d = deserializer;
-        if (d == null) {
+        if (d != null) {
+            return d;
+        } else {
             synchronized (this) {
                 if (deserializer == null) {
-                    d = ctor.get();
-                    deserializer = d;
+                    deserializer = ctor.get();
                 }
             }
+            return deserializer;
         }
-        return d;
     }
 }

--- a/java-client/src/test/java/co/elastic/clients/json/LazyDeserializerTest.java
+++ b/java-client/src/test/java/co/elastic/clients/json/LazyDeserializerTest.java
@@ -1,0 +1,69 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package co.elastic.clients.json;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.concurrent.CompletableFuture;
+
+public class LazyDeserializerTest extends Assert {
+
+
+    @Test
+    public void testConcurrentInit() throws Exception {
+        // See https://github.com/elastic/elasticsearch-java/issues/58
+
+        final LazyDeserializer<String> ld = new LazyDeserializer<>(JsonpDeserializer::stringDeserializer);
+
+        CompletableFuture<JsonpDeserializer<String>> fut1;
+        CompletableFuture<JsonpDeserializer<String>> fut2;
+
+        // Lock the mutex and start 2 threads that will compete for it.
+        synchronized (ld) {
+            fut1 = futureUnwrap(ld);
+            fut2 = futureUnwrap(ld);
+        }
+
+        // We should see the same non-null results everywhere
+        assertNotNull(fut1.get());
+        assertNotNull(fut2.get());
+
+        final JsonpDeserializer<String> unwrapped = ld.unwrap();
+        assertEquals(unwrapped, fut1.get());
+        assertEquals(unwrapped, fut2.get());
+
+    }
+
+    private CompletableFuture<JsonpDeserializer<String>> futureUnwrap(LazyDeserializer<String> d) {
+
+        final CompletableFuture<JsonpDeserializer<String>> result = new CompletableFuture<>();
+
+        new Thread(() -> {
+            try {
+                result.complete(d.unwrap());
+            } catch (Throwable e) {
+                result.completeExceptionally(e);
+            }
+        }).start();
+
+        return result;
+    }
+}

